src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6455,10 +6455,19 @@ static NullableFlowState getReferenceConversionResultState(TypeWithAnnotations t
                 switch (state)
                 {
                     case NullableFlowState.MaybeNull:
-                        if (operandType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() != true &&
-                            targetType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
+                        if (targetType.Type?.IsTypeParameterDisallowingAnnotationInCSharp8() == true)
                         {
-                            return NullableFlowState.MaybeDefault;
+                            var type = operandType.Type;
+                            if (type?.IsTypeParameterDisallowingAnnotationInCSharp8() != true)
+                            {
+                                return NullableFlowState.MaybeDefault;
+                            }
+                            else if (targetType.NullableAnnotation.IsNotAnnotated() &&
+                                type is TypeParameterSymbol typeParameter1 &&
+                                dependsOnTypeParameter(typeParameter1, (TypeParameterSymbol)targetType.Type, NullableAnnotation.NotAnnotated, out var annotation))
```

Comment:
can this cast fail? consider doing `targetType.Type is TypeParameterSymbol typeParameter2 &&` as in the following local function

Reply:
The cast cannot fail since we've checked `IsTypeParameterDisallowingAnnotationInCSharp8()` above.

---
In reply to: [466687268](https://github.com/dotnet/roslyn/pull/46405#discussion_r466687268) [](ancestors = 466687268)